### ISSUE TITLE
Magic & other custom skills

### ISSUE THREAD

**Issue by MichaelMackus** [LambdaHack/LambdaHack#233]
It looks to me like magic (as an actor skill) is unsupported. It would be good to have customizable skill sets for actors, since some/most games will want skills other than those that are already built into the engine. For example, for a "Thief" class he could have skills like "move silently", "hide", "climbing" and other things that wouldn't be available to others.

**Comment by MichaelMackus** [comment]
Added a [wiki page](https://github.com/LambdaHack/LambdaHack/wiki/Magic-&-customizable-skills) with more fleshed out ideas.

**Comment by MichaelMackus** [comment]
Related #234

**Comment by Mikolaj** [comment]
Yes, it makes sense. I think there are several potential useful feature here:

1. Various new fun effects need to be implemented in the engine, e.g., those mentioned above and on the wiki. Definitely, except Detect Traps and Listen that are already there (but can be improved, for sure).

2. Each actor needs a subset of all possible effects engine makes possible. This is already done. The subset is determined by equipment and organs.

3. The subsets needs to be customized, determined, changed via methods X, Y and Z. If X, Y and Z can't be reduced to equipment and organs, we can think of extending the engine to enable them.

4. Classes need to be added. Probably can be via content. Try, I will guide you and if we get stuck we will extend the engine.

5. The game needs custom stats (as the first of `Skills` but defined in content), such as magic level, mana, piety, standing with various deities, and generally anything that can be referred to in other content, in particular as precondition or fuel for some listed effects.

6. 'custom user-defined skills', which in my jargon would probably mean user-defined effects. I'm afraid this is very complex. In full generality this would mean letting content inject arbitrary code that mucks around with the internal state of the game. I think this defeats the purpose of the separation between content and engine. Instead, the game creator should extend the engine as required. It's for this reason that the engine is also a library and all parts can be overridden from the outside though it's often more convenient to rewrite their source code (and later on contribute it back). OTOH, in restricted generality, complex effects can already be programmed using simple programming constructs (composition, partially conditional) from basic effects. Both the set of constructs and of the basic effects can be extended (by orders of magnitude) to let people do a lot of common things and then compose them into less common variants. Ideally, this would be quite orthogonal sets.

Anyway, I guess we need to start somewhere, instead of tackling all 6 features. Which one would it be? Could you open an issue for it and leave this as an umbrella epic?

**Comment by MichaelMackus** [comment]
I guess what I had in mind was more of a higher level abstraction that allowed for custom user-defined skills. So, perhaps a new data kind that abstracts the skill and its progression for actor(s). I really don't feel like its going to be very extensible to keep adding baked-in functionality for this kind of stuff, especially since things like "magic" and "class" are very personal to the game implementation and can change quite drastically from game to game. For example, I might want a vancian-based magic system in one game with a separation between arcane & divine magic, and then in another use some sort of mana pool with different magic disciplines using the same logic for each discipline.

Another use-case would be a skill system such as the one in [Cataclysm: DDA](https://github.com/CleverRaven/Cataclysm-DDA). This is a post-apocalyptic roguelike with a very in-depth crafting system, which allows recipes to be unlocked through in-game learning via books based on the user's current knowledge of the particular skill. I feel like the LambdaHack engine would be suited *very well* to a CDDA-type game except for the fact the skill system is extremely specific for the game in question.

**Comment by MichaelMackus** [comment]
Sure, I can try that as an experiment. Any pointers would be appreciated -
when I looked into it before I was at a loss how this would be done via
game content.

**Comment by Mikolaj** [comment]
Splendid. So what exactly, of the above, would you like to implement? Or is it all of the above in one game, for a start?

**Comment by Mikolaj** [comment]
BTW, wouldn't it be easier to brainstorm on LambdaHack room on Haskell GameDev Discord https://discord.gg/87Ghnws or its mirror room on Matrix (guests are welcome without logging in, but I'm not sure it's read-write) https://matrix.to/#/!HnbpAHMjOGHlYtrASl:mozilla.org?via=mozilla.org&via=t2bot.io&via=matrix.org
